addons/l10n_in/models/account_invoice.py

```diff
@@ -50,17 +50,26 @@ def _compute_l10n_in_gst_treatment(self):
                 record.l10n_in_gst_treatment = gst_treatment
         (self - indian_invoice).l10n_in_gst_treatment = False
 
+    def _get_l10n_in_pos_sale_partner(self):
+        self.ensure_one()
+        if (self.partner_id.commercial_partner_id == self.partner_shipping_id.commercial_partner_id
+            and self.partner_shipping_id.state_id
+        ):
+            return self.partner_shipping_id
+        elif self.partner_id.state_id:
+            return self.partner_id
+        elif self.partner_id.commercial_partner_id.state_id:
+            return self.partner_id.commercial_partner_id
+        return self.env['res.partner'].browse()
+
     @api.depends('partner_id', 'partner_shipping_id', 'company_id')
     def _compute_l10n_in_state_id(self):
         for move in self:
             if move.country_code == 'IN' and move.journal_id.type == 'sale':
-                partner_state = (
-                    move.partner_id.commercial_partner_id == move.partner_shipping_id.commercial_partner_id
-                    and move.partner_shipping_id.state_id
-                    or move.partner_id.state_id
-                )
-                if not partner_state:
-                    partner_state = move.partner_id.commercial_partner_id.state_id or move.company_id.state_id
+                if partner := move._get_l10n_in_pos_sale_partner():
```

Comment:
This should work fine :smile: 
```suggestion
         partner = move._get_l10n_in_pos_sale_partner()
         partner_state = partner.state_id or move.company_id.state_id
```